src/Geocoder/Provider/AbstractHttpProvider.php

```diff
@@ -18,27 +21,33 @@
 class AbstractHttpProvider extends AbstractProvider
 {
     /**
-     * @var HttpAdapterInterface
+     * @var HttpClient
      */
-    private $adapter;
+    protected $client;
 
     /**
-     * @param HttpAdapterInterface $adapter An HTTP adapter
+     * @var MessageFactory
      */
-    public function __construct(HttpAdapterInterface $adapter)
+    protected $messageFactory;
+
+    /**
+     * @param HttpClient $client An HTTP adapter
+     */
+    public function __construct(HttpClient $client = null, MessageFactory $factory = null)
     {
         parent::__construct();
 
-        $this->adapter = $adapter;
+        $this->client = $client ?: HttpClientDiscovery::find();
+        $this->messageFactory = $factory ?: MessageFactoryDiscovery::find();
```

Comment:
Put these in the getters. This will allow us not to use puli discovery by using `$provider->setHttpMessageFactory($factory)`